kratos/utilities/parallel_utilities.h

```diff
@@ -0,0 +1,364 @@
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//                   Denis Demidov
+//
+
+#if !defined(KRATOS_PARALLEL_UTILITIES_H_INCLUDED )
+#define  KRATOS_PARALLEL_UTILITIES_H_INCLUDED
+
+
+// System includes
+#include<iostream>
+#include<array>
+#include<vector>
+#include<tuple>
+#include<cmath>
+#include<limits>
+#include<omp.h>
+
+#include <future>
+#include <thread>
+
+// External includes
+
+
+// Project includes
+#include "includes/define.h"
+
+
+namespace Kratos
+{
+///@addtogroup KratosCore
+
+//***********************************************************************************
+//***********************************************************************************
+//***********************************************************************************
+template<class TContainerType, class TIteratorType=typename TContainerType::iterator>
+class BlockPartition
+{
+public:
+
+    static constexpr int max_allowed_threads = 128;
+
+    BlockPartition(TIteratorType it_begin,
+                   TIteratorType it_end,
+                   int Nchunks = omp_get_max_threads())
+        : mit_begin(it_begin), mit_end(it_end), mNchunks(Nchunks)
+    {
+        int NumTerms = it_end-it_begin;
+        int mBlockPartitionize = NumTerms / mNchunks;
+        mBlockPartition[0] = 0;
+        mBlockPartition[mNchunks] = NumTerms;
+        for(int i = 1; i < mNchunks; i++)
+            mBlockPartition[i] = mBlockPartition[i-1] + mBlockPartitionize ;
+
+    }
+
+    BlockPartition(TContainerType& rData,
+                   int Nchunks = omp_get_max_threads())
+        : BlockPartition(rData.begin(), rData.end(), Nchunks)
+    {}
+
+    virtual ~BlockPartition() {}
+
+    //version enabling reduction
+    //simple version
+    template <class TUnaryFunction>
+    inline void for_each(TUnaryFunction&& f)
+    {
+        #pragma omp parallel for
+        for(int i=0; i<mNchunks; ++i)
+        {
+            for (int k = mBlockPartition[i]; k < mBlockPartition[i+1]; ++k)
+            {
+                f(*(mit_begin + k)); //note that we pass a reference to the value, not the iterator
+            }
+        }
+    }
+
+    //version with reduction
+    template <class TReducer, class TUnaryFunction>
+    inline TReducer for_reduce(TUnaryFunction &&f)
+    {
+        TReducer global_reducer;
+        #pragma omp parallel for
+        for(int i=0; i<mNchunks; ++i)
+        {
+            TReducer local_reducer;
+            for (auto k = mBlockPartition[i]; k < mBlockPartition[i+1]; ++k)
+            {
+                local_reducer.LocalMerge(f(*(mit_begin + k)));
+            }
+            global_reducer.ThreadSafeMerge(local_reducer);
+        }
+        return global_reducer;
+    }
+
+private:
+
+    TIteratorType mit_begin, mit_end;
+    int mNchunks;
+    std::array<int, max_allowed_threads> mBlockPartition;
```

Comment:
`int` here limits the size of vectors that can be processed by 2e9 elements. Is that acceptable for Kratos?

Reply:
again a good point. The thing is that if our backend is OpenMP than IN WINDOWS we are compelled to use int

Reply:
I think windows allows `ptrddiff_t` as well. It just wants the index to be signed.

Reply:
Also, you do `omp for` in the outer loop (over mNumChunks). Inner loop, where you use mBlockPartition, is not openmp.

Reply:
ok, i will change this.